dotCMS/src/main/java/com/dotmarketing/services/PageServices.java

```diff
@@ -197,126 +197,142 @@ public static InputStream buildStream(IHTMLPage htmlPage, Identifier identifier,
-			}
-			if(contentletsFull.size() > 0){
-                Set<String> contentletIdentList = new HashSet<String>();
-                List<Contentlet> contentletsFilter = new ArrayList<Contentlet>();
-                for(Contentlet cont : contentletsFull){
-                    if(!contentletIdentList.contains(cont.getIdentifier())){
-                        contentletIdentList.add(cont.getIdentifier());
-                        contentletsFilter.add(cont);
-                    }
-                }
-                contentletsFull = contentletsFilter;
-            }
-			
-			StringBuilder widgetpree=new StringBuilder();
-			StringBuilder widgetpreeFull=new StringBuilder();
-			
-			StringBuilder contentletList = new StringBuilder();
-			int count=0;
-			for(Contentlet contentlet : contentlets) {
-			    contentletList.append(count==0 ? "" : ",")
-			        .append('"').append(contentlet.getIdentifier()).append('"');
-			    if(contentlet.getStructure().getStructureType()== Structure.STRUCTURE_TYPE_WIDGET) {
-                    Field field=contentlet.getStructure().getFieldVar("widgetPreexecute");
-                    if (field!= null && UtilMethods.isSet(field.getValues()))
-                        widgetpree.append(field.getValues().trim());
-                }
-			    if(++count>=c.getMaxContentlets()) break;
-			}
-			
-			StringBuilder contentletListFull = new StringBuilder();
-            int countFull=0;
-            for(Contentlet contentlet : contentletsFull) {
-                contentletListFull.append(countFull==0 ? "" : ",")
-                    .append('"').append(contentlet.getIdentifier()).append('"');
-                if(contentlet.getStructure().getStructureType()== Structure.STRUCTURE_TYPE_WIDGET) {
-                    Field field=contentlet.getStructure().getFieldVar("widgetPreexecute");
-                    if (field!= null && UtilMethods.isSet(field.getValues()))
-                        widgetpreeFull.append(field.getValues().trim());
-                }
-                if(++countFull>=c.getMaxContentlets()) break;
-            }
 
-			//Check if we want to accrue the tags associated to each contentlet on this page
-			if ( Config.getBooleanProperty("ACCRUE_TAGS_IN_CONTENTS_ON_PAGE", false) ) {
+				//Check if we want to accrue the tags associated to each contentlet on this page
+				if (Config.getBooleanProperty("ACCRUE_TAGS_IN_CONTENTS_ON_PAGE", false)) {
 
-				for ( Contentlet contentlet : contentletsFull ) {
-					//Search for the tags associated to this contentlet inode
-					List<Tag> contentletFoundTags = APILocator.getTagAPI().getTagsByInode(contentlet.getInode());
-					if ( contentletFoundTags != null ) {
-						pageFoundTags.addAll(contentletFoundTags);
+					for (Contentlet contentlet : contentletsFull) {
+						//Search for the tags associated to this contentlet inode
+						List<Tag> contentletFoundTags = APILocator.getTagAPI().getTagsByInode(contentlet.getInode());
+						if (contentletFoundTags != null) {
```

Comment:
![CRITICAL](https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL') Refactor this code to not nest more than 4 if/for/while/switch/try statements. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS134)